core/src/main/java/org/testcontainers/dockerclient/DockerClientProviderStrategy.java

```diff
@@ -114,108 +115,33 @@ public static DockerClientProviderStrategy getFirstValidStrategy(List<DockerClie
-                        log.debug(
-                            "Transport type: '{}', Docker host: '{}'",
-                            TestcontainersConfiguration.getInstance().getTransportType(),
-                            strategy.getTransportConfig().getDockerHost()
-                        );
-
-                        log.debug("Checking Docker OS type for {}", strategy.getDescription());
-                        String osType = info.getOsType();
-                        if (StringUtils.isBlank(osType)) {
-                            log.warn("Could not determine Docker OS type");
-                        } else if (!osType.equals("linux")) {
-                            log.warn("{} is currently not supported", osType);
-                            throw new InvalidConfigurationException(osType + " containers are currently not supported");
-                        }
-
-                        if (strategy.isPersistable()) {
-                            TestcontainersConfiguration.getInstance().updateUserConfig("docker.client.strategy", strategy.getClass().getName());
-                        }
-
-                        return Stream.of(strategy);
-                    } catch (Exception | ExceptionInInitializerError | NoClassDefFoundError e) {
-                        @Nullable String throwableMessage = e.getMessage();
-                        @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
-                        Throwable rootCause = Throwables.getRootCause(e);
-                        @Nullable String rootCauseMessage = rootCause.getMessage();
-
-                        String failureDescription;
-                        if (throwableMessage != null && throwableMessage.equals(rootCauseMessage)) {
-                            failureDescription = String.format("%s: failed with exception %s (%s)",
-                                    strategy.getClass().getSimpleName(),
-                                    e.getClass().getSimpleName(),
-                                    throwableMessage);
-                        } else {
-                            failureDescription = String.format("%s: failed with exception %s (%s). Root cause %s (%s)",
-                                    strategy.getClass().getSimpleName(),
-                                    e.getClass().getSimpleName(),
-                                    throwableMessage,
-                                    rootCause.getClass().getSimpleName(),
-                                    rootCauseMessage
-                            );
-                        }
-                        configurationFailures.add(failureDescription);
-
-                        log.debug(failureDescription);
-                        return Stream.empty();
-                    }
-                })
+                .flatMap(strategy -> tryOutStrategy(configurationFailures, strategy))
                 .findAny()
```

Comment:
Sorry: pedantic comment, and I think this comes from the original. Surely this should be `findFirst()`?

I don't think it's terribly risky, but again it's slightly clearer about the intent.

There's a separate can of ~worms~opportunities that also opens up when considering that we could try strategies in parallel (but just select the highest priority one that succeeds). I'd not push you to explore that in this PR though :)

Reply:
Not pedantic at all, but again, that is how the original code does it and I consciously did not touch it.
And I think you are right, this is semantically wrong, since it will not comply with our priorities (although, since we run the stream non-parallel, it is most likely equivalent to `findFirst()`, but strictly speaking not guaranteed 😨 ).

I am happy to change at least the method in this PR.